### ISSUE TITLE
linear interpolation leads to some crazy estimates

### ISSUE THREAD

**Issue by astolman** [dialtr/libcount#14]
The linear interpolation you wrote for estimating the biases seems like a reasonable approach, but I found a case where it leads to absurd estimates. I had a counter of precision 5 with a raw estimate of 128 point 6 something and the two closest raw estimates libcount found in the table were like 128.3464 and 128.3462, and their respective biases are about 1 apart. The linear interpolation said that it's bias should be in the thousands since the straight line connecting those two closest points is so steep. As a result, I got back a negative cardinality estimation, which is definitely not right. The spec in the paper just says "interpolate" without specifying how, so I would instead of just looking for the two nearest neighbors, look for the interval which contains the raw estimate and find where the estimate lies on that line. That seems like another reasonable interpolation and avoids the bias estimate for a given value to be wildly larger than that of its nearest neighbors.

**Comment by dialtr** [comment]
That sounds very reasonable. I'll work on a fix.

**Comment by dialtr** [comment]
Hey Andrew, I finally got around to a fix for this, would love to get your review on it. The branch is:

   tdial-interpolation-fix

I am working on certification tests next.

**Comment by dialtr** [comment]
Actually, hold off. There's an issue with the data in the tables.

**Comment by astolman** [comment]
I wrote a fix for it in my adaptation of libcount. I think I had to sort the data that you used because it wasn’t sorted in a few cases. You can see what I did in my project https://github.com/astolman/HyperHeadTail <https://github.com/astolman/HyperHeadTail> in the shll folder there. I needed a HyperLogLog implementation that supported a sliding window so I heavily rewrote the hll.cc files, but the rest of it is changed as little as possible. It’s been a while, but I think the stuff in empirical_data.cc contains the now appropriately sorted values.

**Comment by dialtr** [comment]
Yep, the original data was definitely not sorted. I had contacted one of
the original authors about that (Marc) and he added my observation to the
fine print on the Google doc.  This is ultimately why I elected to use
nearest neighbor instead of sorting it (and then got it wrong!)

I've subsequently decided to take your suggestion and sort it: the simple
linear interpolation that you suggested is *much* easier to do than
anything more sophisticated.

I'll take a look at what you have.

I am going to clean this up further and simplify it where possible, as well
as provide some certification tests.  Since it sounds like you've worked
with this a little, do you have any suggestions in that direction?

Also, I am delighted that you found use for the code, even if I am a little
embarrassed with the bugs - if you have time to explain what you're doing
with it, I'd be interested to know!

Warm Regards,

TD

**Comment by dialtr** [comment]
P.S.

Unless I am mistaken, in your version of EMP_alpha(), I think you have the
first three switch cases(16, 32, 64) wrong.

Or am I missing something?

TD

**Comment by astolman** [comment]
I think these cases work out so that your cases are log of mine. I think I did that as part of fixing the problem with the first bug report I submitted. 

As far as my use for libcount, I did a summer project at Sandia National Labs where I used it for a streaming algorithm. I’m still (slowly) writing a paper for it that should be up on ArXiv soon.

**Comment by dialtr** [comment]
Linear interpolation fix completed. Note that the empirical data is now sorted properly. Also added a simple certification test.  Just run:

  make certify
